### PR TITLE
feat: add typescript client postmessage support

### PR DESCRIPTION
**Pull request by shanejonas**
image

### CODE REVIEW

File: src/components/client.ts [hunk 1]
```diff
@@ -38,18 +40,21 @@ export class <%= className %> {
...
+    const {type, host, port, protocol, postMessageType} = options.transport;
     let path = options.transport.path || "";
     if(path && path[0] !== "/") {
         path = "/" + path;
     }
     switch (type) {
       case 'http':
       case 'https':
-        this.transport = new HTTPTransport(type + "://" + host + ":" + port + path)
+        this.transport = new HTTPTransport((protocol || type) + "://" + host + ":" + port + path);
```
Comment (zcstarr) [inline src/components/client.ts:51]: why protocol here?

File: templates/client/typescript/_package.json [hunk 2]
```diff
@@ -13,13 +13,13 @@
   "author": "",
   "license": "Apache-2.0",
   "dependencies": {
-    "@open-rpc/client-js": "^1.2.3",
+    "@open-rpc/client-js": "^1.3.0",
     "@open-rpc/meta-schema": "^1.6.0",
     "@open-rpc/schema-utils-js": "^1.12.0",
-    "lodash": "^4.17.15",
     "@types/json-schema": "7.0.3",
+    "@types/lodash": "^4.14.149",
```
Comment (BelfordZ) [inline templates/client/typescript/_package.json:20]: Should be dev dep

File: src/components/client.ts (hunk 1 above)
Comment (zcstarr) [reply src/components/client.ts:50]: I think we should get rid of protocol and type and create a mapping structure. Then you can change the switch on type to simply retrieve the translation from a map. something like 

```ts
type ProtocolNames = "ws" | "http" | "https" | "postmessageiframe" | "postmessagewindow";
type NameMapping = Record<ProtocolNames,string>
const protocolNaming: NameMapping = {
  "ws": "websocket",
  "http":"http",
  "https": "https",
  "postmessageiframe": "postmessage",
  "postmessagewindow": "postmessagew"
}
const protocolType = "ws"
protocolNaming[protocolType]
```

File: src/components/client.ts [hunk 3]
```diff
@@ -19,17 +19,18 @@ import { MethodCallValidator, MethodNotFoundError } from "@open-rpc/schema-utils
 
 export interface Options {
   transport: {
-    type: "websocket" | "http" | "https";
+    type: "websocket" | "http" | "https" | "postmessagewindow" | "postmessageiframe";
     host: string;
     port: number;
     path?: string;
-  }
+    protocol?: string;
```
Comment (zcstarr) [inline src/components/client.ts:26]: see comment about protocol. It seems like we could lose this parameter and just do a logical name mapping internally from type => protocol names. Unless protocol is used in some other way.

File: src/components/client.ts (hunk 1 above)
Comment (shanejonas) [reply src/components/client.ts:50]: I want the user to be able to override protocol

Comment (shanejonas) [reply src/components/client.ts:50]: `http://localhost:3000` vs `https://sig.tools`


there is no way to define this any other way
image

Comment (shanejonas) [reply src/components/client.ts:50]: using `protocol` here would follow the nodejs URL parsing API of `.protocol`: https://nodejs.org/api/url.html#url_url_protocol

**Pull request merged**